tests: add asynchronous log browsing functionality

margdoc opened the pull request:
# Motivation

To write certain tests, we need the ability to asynchronously browse node logs inside tests.
For example:
* it can be used to detect segfaults. Reproducer https://github.com/margdoc/scylla/commit/97d6946e314a5b5d445b57045d93c57d10494b48 (issue https://github.com/scylladb/scylladb/issues/14397) could read the log to determine whether there was a segmentation fault or a graceful shutdown of the node.
* wait for some events that happen in the background, for example in the dtest `bootstrap_test.py::test_decommissioned_wiped_node_can_join`, the main thread waits for the first node to notice that the fourth node is down.
* check if we print appropriate errors and warnings to the user. For example tests in the file `cql-pytest/test_logs.py` in test.py (but it has only synchronous wait_for).



# Features
Add a class that handles log file browsing with the following features:
* mark: returns "a mark" to the current position of the log.
* wait_for: asynchronously checks if the log contains the given message.
* grep: returns a list of lines matching the regular expression in the log.

Add a new endpoint in `ManagerClient` to obtain the Scylla logfile path.

# Examples

Real test example: https://github.com/margdoc/scylla/commit/a227cc9ea63e013be019ac09bcc0845d42b00fc4
This is the reproducer for https://github.com/scylladb/scylladb/issues/14397. It uses this feature to enforce a specific order of events.
In the other reproducer of the same issue (https://github.com/margdoc/scylla/commit/97d6946e314a5b5d445b57045d93c57d10494b48) we could read the scylla log to check if there was a segmentation fault or if the node stopped gracefully.

--------

Fixes #14782.

nyh commented:
The commit message says "add functionality", but in fact you're moving good, working, functionality out of cql-pytest.

The question is why.

Right now cql-pytest doesn't depend on anything in pylib. To be honest I am worried about the bloat in test/pylib, I'm worried about unnecessary complications it encourages (like typing and asynchronous programming) and don't trust that as it evolves, it will remain suitable for cql-pytest and not start needing more dependencies, both internal (additional convenience functions, etc.) or external (more libraries, etc.).

Why are we so afraid of code duplication?

nyh reviewed (changes requested):
Something I repeatedly keep asking for (and documented officially in test/cql-pytest/README.md) is to focus on writing tests, not a test library. At least in the test/cql-pytest directory (you can do other things in other test/ subdirectories). New test framework features should be added sparingly, and motivated by the real needs of real tests.

I assume these changes had some motivation of writing a real test, perhaps in the test/topology framework. If this is the case, why doesn't this PR include such a test? Why does the main test added in this PR test the test framework, and not Scylla?

Because of this, I prefer that if you'd like to do these things, that you don't touch the cql-pytest directory at all and just copy the small amount of code (just 60 lines for local_process_id() and logfile_path() - you didn't reuse anything else) into test/pylib, and then continue to add to it you new class-based APIs you wanted (while cql-pytest can continue to use the wait_for_log() function it already had).

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2700/

nyh commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2700/

The patch broke the rest_api tests:

```
ImportError while loading conftest '/jenkins/workspace/scylla-master/scylla-ci/scylla/test/rest_api/conftest.py'.
test/rest_api/conftest.py:23: in <module>
    from util import unique_name
test/cql-pytest/util.py:23: in <module>
    from test.pylib.log_browsing import node_process_id
E   ModuleNotFoundError: No module named 'test'
```

This won't be relevant if you don't change cql-pytest, like I proposed above.

margdoc commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,83 @@
+# process exists, return its process id (as a string). Otherwise, return None.
+# Note that the local process needs to belong to the same user running this
+# test, or it cannot be found.
+def node_process_id(server_ip: IPAddress, port: int = 9042) -> Optional[int]: #TODO port
```
> The word "local" in the original name I used for this function, "local_process_id", refers to the fact that it can only find a local process - it can't find a process that runs on some remote machine. I think you shouldn't have lost this word (or something with similar meaning).

Good point. I just wanted to distinguish these two functions. I will change that.

> I don't understand the comment "TODO port" - it seems port is already a parameter?

Sorry, It was a note for me and I deleted it in the later commit. I forgot to delete it from the first one.

margdoc commented:
v2:
* `cql-pytest/test_logs.py` and `pylib/log_browsing.py` are now independent of each other
* made `wait_for` an asynchronous function
* removed the test
* added the real test example in the pr description

kbr-scylla commented:
> Real test example: https://github.com/margdoc/scylla/commit/a227cc9ea63e013be019ac09bcc0845d42b00fc4

Please provide some context.

kbr-scylla commented:
> What's the purpose of the feature?

margdoc commented:
> What's the purpose of the feature?

For example:
* as Kamil mentioned, it can be used to detect segfaults. Reproducer https://github.com/margdoc/scylla/commit/97d6946e314a5b5d445b57045d93c57d10494b48 (issue https://github.com/scylladb/scylladb/issues/14397) could read the log to determine whether there was a segmentation fault or a graceful shutdown of the node.
* wait for some events that happen in the background, for example in the dtest `bootstrap_test.py::test_decommissioned_wiped_node_can_join`, the main thread waits for the first node to notice that the fourth node is down.
* check if we print appropriate errors and warnings to the user. For example tests in the file `cql-pytest/test_logs.py` in test.py.

margdoc commented:
v3:
* add a new endpoint to the ManagerClient.
* scylla log file is obtained from the ManagerClient.

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -116,15 +117,15 @@ def cluster_con(hosts: List[IPAddress], port: int, use_ssl: bool):
 
 @pytest.fixture(scope="session")
-async def manager_internal(event_loop, request):
+async def manager_internal(event_loop, request, thread_pool):
```
Let's move the thread_pool fixture closed to where it's used (i.e. before `manager_internal`).

Even better, since you're storing the thread_pool in ManagerClient now, we shouldn't need a separate fixture at all.
Can we just construct the thread pool inside `manager_internal`?

margdoc replied:
Moved it to the constructor of `ManagerClient`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2984/

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,89 @@
+    # This is for use with the from_mark parameter of watch_log_for method,
+    # allowing to watch the log from the position when this method was called.
+    def mark(self) -> int:
+        with open(self.file, 'r') as f:
```
Shouldn't we do the seeks on the thread pool as well?

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,89 @@
+        with open(self.file) as f:
+            if from_mark:
+                f.seek(from_mark)
+            for line in f:
```
This is accessing the file synchronously.

margdoc replied:
Is it a problem?

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -288,3 +291,8 @@ async def _not_sees_another_server():
                 return True
         await wait_for(_not_sees_another_server, time() + interval, period=.5)
+
+    async def server_get_log_filename(self, server_id: ServerNum) -> ScyllaLogFile:
```
The function is not returning just the file name, it's returning a `ScyllaLogFile`, let's give a better name to the function, maybe just `server_open_log`?

BTW. is there any resource we need to close after obtaining `ScyllaLogFile` and doing whatever we need to?

margdoc replied:
> BTW. is there any resource we need to close after obtaining ScyllaLogFile and doing whatever we need to?

No, we open a log file for each `ScyllaLogFile`'s method call

margdoc commented:
v4:
* made `grep` and `mark` asynchronous functions
* s/server_get_log_filename/server_open_log
* Move the initialization of `thread_pool` to the constructor of `ManagerClient`

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,88 @@
+            async with asyncio.timeout(timeout):
+                logger.debug("Waiting for log message: %s", pattern)
+                while True:
+                    line = await loop.run_in_executor(self.thread_pool, f.readline)
```
`self._run_in_executor`?

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,88 @@
+
+logger = logging.getLogger(__name__)
+
+# Class for browsing a Scylla log file.
```
Please make all the comments Python docstrings.

kbr-scylla replied:
Well, not all, but those that describe the classes/functions at the beginning.

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,88 @@
+            await self._run_in_executor(f.seek, 0, os.SEEK_END)
+            return f.tell()
+
+    # wait_for_log() checks if the log contains the given message.
```
s/wait_for_log/wait_for

(or rename the function to match the comment)

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,88 @@
+        matchings = []
+        pattern = re.compile(expr)
+        if filter_expr:
+            filter_pattern = re.compile(filter_expr)
```
Nit: can be made a bit more succint using ternary operator

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3009/

margdoc commented:
v5:
* fixed the usage of `run_in_executor`
* changed comments to Python docstrings
* minor fixes

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,95 @@
+        if not os.path.isfile(self.file):
+            pytest.fail("Log file {} does not exist".format(self.file))
+
+    async def _run_in_executor(self, func, *args, loop=None):
```
I don't get why you added the `loop` parameter. Isn't it always obtained through `get_running_loop()`? Why sometimes obtain it outside the function and sometimes inside?

margdoc replied:
I wanted to reuse `loop` inside `wait_for` and `grep` to not call `get_running_loop()` on each read line from the log file.
I also wanted to make calling it easier in case when it called only once.

kbr-scylla replied:
Ok

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3016/

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3027/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3031/

avikivity commented:
Why so many CI failures?

The motivation should be part of the cover letter. It's the most important part. Otherwise what grounds would I have to reject a patch that reverts this commit?

nyh commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,95 @@
+                    line = await self._run_in_executor(f.readline, loop=loop)
+                    if line and prog.search(line):
+                        logger.debug("Found log message: %s", line)
+                        return
```
If readline reached EOF, do you try it immediately, again and again, until it starts returning data? Did you test that this actually works (it might not work if Python's stdio library caches the fact it reached EOF!!).
Also, you should probably add a short sleep between retries - no point in doing this loop a million times a second :-(

nyh replied:
I wrote a short test for this and found that:
1. readline() does work after it returned EOF once (it's not cached). So your code seems correct even if not obvious.
2. While waiting, it does a million (!) iterations per second, which is completely excessive. So a sleep, even a very short one (e.g., 0.01 seconds) is probably better than nothing.

nyh commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,95 @@
+            async with asyncio.timeout(timeout):
+                logger.debug("Waiting for log message: %s", pattern)
+                while True:
+                    line = await self._run_in_executor(f.readline, loop=loop)
```
Are you sure that the log file is line buffered (half-written lines are never readable)?

If this assumption is not true (maybe it is!), then this code is not correct, and to fix it you need to do the following: id if readline reads until EOF a line that doesn't end in a newline, it needs to be saved in memory and the following readline which will succeed later will need to be appended to it.

nyh reviewed (commented):
Looks good to me except some comments I left in.

In general I'm not happy about adding test machinary before actual tests need them. Doing that can result in a utility function that isn't actually useful for real tests, and may even contain bugs. But since this is not cql-pytest but test/pylib I know that @kostja and @kbr-scylla have different ideals than I do so I'll let them decide.

kbr- commented:
@nyh 

> In general I'm not happy about adding test machinary before actual tests need them.

https://github.com/scylladb/scylladb/pull/15004

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3075/

margdoc commented:
v6:
* `wait_for`: add a sleep in the loop
* `wait_for`: add line buffering

margdoc commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,95 @@
+        """
+        with open(self.file, 'r') as f:
+            await self._run_in_executor(f.seek, 0, os.SEEK_END)
+            return await self._run_in_executor(f.tell)
```
In my opinion it isn't a big difference, it's just a short IO operation. Kamil had an objection to the first version where it was synchronous, so I changed that.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3103/

nyh commented on test/pylib/log_browsing.py:
```diff
@@ -0,0 +1,101 @@
+            async with asyncio.timeout(timeout):
+                logger.debug("Waiting for log message: %s", pattern)
+                while True:
+                    await asyncio.sleep(0.01)
```
You put the sleep at the wrong place... you should sleep in an else of "if line" below. In other words, sleep a bit when we reached the end of file. As the code is written here, it will sleep between each line! In other words, it will be limited to reading 100 log lines per second.

margdoc commented:
v7:
* moved `sleep`
* rebased on master

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3264/
